### PR TITLE
Cbecker

### PR DESCRIPTION
**Pull request by charlie-becker**
I found a bug in the categorial model as it was not updating the epoch correctly for the annealing coefficient and as a result I could not get similar results for a model trained with the deprecated ( keras 2) version. This PR fixes that.

Additionally, I did some refactoring that abstracts the callbacks and custom loss functions away to make it easier for the end user to use. It also now can easily used as a base class for regular regression and classification, and all that is needed for the evidential approach is to set the kwarg `evidential=True` (along with a coefficient if not using the default). The regression model also supports the Gaussian parametric uncertainty approach with the kwarg `uncertainty=True`. Note, if both `evidential` and `uncertainty` are set to `True`, the evidential model will override. The regression model also supports an arbitrary number of output tasks. I have included tests for all the use cases I just listed. 

Here's some sample code (model compilation, etc is under the hood):

**Evidential Categorical:**
```
 model = CategoricalDNN(evidential=True,
                                            n_classes=n_classes,
                                            n_inputs=n_features,
                                            epochs=10,
                                            annealing_coeff=1.5,
                                            lr=0.0001)
        hist = model.fit(x_train, y_train)
        p_with_uncertainty = model.predict(x_train, return_uncertainties=True, batch_size=10000)
```

**Evidential Regression:**
```
 model = RegressorDNN(hidden_layers=2,
                                          n_output_tasks=n_outputs,
                                          n_inputs=n_features,
                                          evidential=True,
                                          epochs=10)
            model.fit(x_train, y_train)
            p_with_uncertainty = model.predict(x_train, return_uncertainties=True)
```

**Parametric Gaussian Regression:**
```
model = RegressorDNN(hidden_layers=2,
                                          n_output_tasks=n_outputs,
                                          n_inputs=n_features,
                                          evidential=False,
                                          uncertainty=True,
                                          epochs=10)
            model.fit(x_train, y_train)
            p_with_uncertainty = model.predict(x_train, return_uncertainties=True)
```

### CODE REVIEW

**Review by djgagne — CHANGES_REQUESTED** [review]
The refactoring generally looks good and should increase usability further without requiring too much extra code to maintain. I do have a couple of requests:

1. Can you add your code examples from your pull request comment to the docstrings for each model?
2. In newer versions of keras, you can access the Leaky ReLU with the `Activation("leaky_relu")` layer, so the custom LeakyReLU layer and the associated if statements are no longer needed.  That may be why you couldn't get it to work earlier.

**Comment by charlie-becker** [comment]
Thanks for the review. Yes, the leaky relu issue was simply solved by using the "leaky_relu" string, and I have added the documentation examples to the doc strings.

**Comment by charlie-becker** [comment]
@djgagne Is this ready to merge and update to pypi?

**Review by djgagne — APPROVED** [review]
Looks good. Merging.
